File: sdk/packages/node/helpers/src/observability/otel-worker-gauges.ts
import type { Meter, BatchObservableResult, Observable } from '@opentelemetry/api'
import { WorkerMetricsCollector } from './worker-metrics'

export interface WorkerGaugesOptions {
  workerId: string
  workerName?: string
}

let registeredGauges = false
let metricsCollector: WorkerMetricsCollector | null = null
let registeredMeter: Meter | null = null
let registeredBatchCallback: ((observableResult: BatchObservableResult) => void) | null = null
let registeredObservables: Observable[] = []

export function registerWorkerGauges(meter: Meter, options: WorkerGaugesOptions): void {
  if (registeredGauges) {
    return
  }

  const { workerId, workerName } = options
  const baseAttributes = {
    'worker.id': workerId,
    ...(workerName && { 'worker.name': workerName }),
  }

  metricsCollector = new WorkerMetricsCollector()

  const memoryHeapUsed = meter.createObservableGauge('iii.worker.memory.heap_used', {
    description: 'Worker heap memory used in bytes',
    unit: 'bytes',
  })

  const memoryHeapTotal = meter.createObservableGauge('iii.worker.memory.heap_total', {
    description: 'Worker total heap memory in bytes',
    unit: 'bytes',
  })

  const memoryRss = meter.createObservableGauge('iii.worker.memory.rss', {
    description: 'Worker resident set size in bytes',
    unit: 'bytes',
  })

  const memoryExternal = meter.createObservableGauge('iii.worker.memory.external', {
    description: 'Worker external memory in bytes',
    unit: 'bytes',
  })

  const cpuPercent = meter.createObservableGauge('iii.worker.cpu.percent', {
    description: 'Worker CPU usage percentage',
    unit: '%',
  })

  const cpuUserMicros = meter.createObservableGauge('iii.worker.cpu.user_micros', {
    description: 'Worker CPU user time in microseconds',
    unit: 'us',
  })

  const cpuSystemMicros = meter.createObservableGauge('iii.worker.cpu.system_micros', {
    description: 'Worker CPU system time in microseconds',
    unit: 'us',
  })

  const eventLoopLag = meter.createObservableGauge('iii.worker.event_loop.lag_ms', {
    description: 'Worker event loop lag in milliseconds',
    unit: 'ms',
  })

  const uptimeSeconds = meter.createObservableGauge('iii.worker.uptime_seconds', {
    description: 'Worker uptime in seconds',
    unit: 's',
  })

  const batchCallback = (observableResult: BatchObservableResult) => {
    if (!metricsCollector) return

    const metrics = metricsCollector.collect()

    if (metrics.memory_heap_used !== undefined) {
      observableResult.observe(memoryHeapUsed, metrics.memory_heap_used, baseAttributes)
    }
    if (metrics.memory_heap_total !== undefined) {
      observableResult.observe(memoryHeapTotal, metrics.memory_heap_total, baseAttributes)
    }
    if (metrics.memory_rss !== undefined) {
      observableResult.observe(memoryRss, metrics.memory_rss, baseAttributes)
    }
    if (metrics.memory_external !== undefined) {
      observableResult.observe(memoryExternal, metrics.memory_external, baseAttributes)
    }
    if (metrics.cpu_percent !== undefined) {
      observableResult.observe(cpuPercent, metrics.cpu_percent, baseAttributes)
    }
    if (metrics.cpu_user_micros !== undefined) {
      observableResult.observe(cpuUserMicros, metrics.cpu_user_micros, baseAttributes)
    }
    if (metrics.cpu_system_micros !== undefined) {
      observableResult.observe(cpuSystemMicros, metrics.cpu_system_micros, baseAttributes)
    }
    if (metrics.event_loop_lag_ms !== undefined) {
      observableResult.observe(eventLoopLag, metrics.event_loop_lag_ms, baseAttributes)
    }
    if (metrics.uptime_seconds !== undefined) {
      observableResult.observe(uptimeSeconds, metrics.uptime_seconds, baseAttributes)
    }
  }

  meter.addBatchObservableCallback(batchCallback, [
    memoryHeapUsed,
    memoryHeapTotal,
    memoryRss,
    memoryExternal,
    cpuPercent,
    cpuUserMicros,
    cpuSystemMicros,
    eventLoopLag,
    uptimeSeconds,
  ])

  registeredMeter = meter
  registeredBatchCallback = batchCallback
  registeredObservables = [
    memoryHeapUsed,
    memoryHeapTotal,
    memoryRss,
    memoryExternal,
    cpuPercent,
    cpuUserMicros,
    cpuSystemMicros,
    eventLoopLag,
    uptimeSeconds,
  ]

  registeredGauges = true
}

export function stopWorkerGauges(): void {
  // Remove the batch observable callback before stopping
  if (registeredMeter && registeredBatchCallback) {
    registeredMeter.removeBatchObservableCallback(registeredBatchCallback, registeredObservables)
  }

  if (metricsCollector) {
    metricsCollector.stopMonitoring()
    metricsCollector = null
  }

  registeredMeter = null
  registeredBatchCallback = null
  registeredObservables = []
  registeredGauges = false
}

File: sdk/packages/rust/helpers/src/observability/telemetry/otel_worker_gauges.rs
use opentelemetry::KeyValue;
use opentelemetry::metrics::Meter;
use std::sync::Arc;

use super::worker_metrics::WorkerMetricsCollector;

/// Options for registering worker gauges
pub struct WorkerGaugesOptions {
    pub worker_id: String,
    pub worker_name: Option<String>,
}

/// Register observable gauges for worker metrics with the given meter.
///
/// Returns a handle that keeps the gauges alive. Drop it to stop reporting.
pub fn register_worker_gauges(
    meter: &Meter,
    collector: Arc<WorkerMetricsCollector>,
    options: WorkerGaugesOptions,
) -> WorkerGaugesHandle {
    let mut attrs = vec![KeyValue::new("worker.id", options.worker_id)];
    if let Some(name) = options.worker_name {
        attrs.push(KeyValue::new("worker.name", name));
    }
    let attrs = Arc::new(attrs);

    let c = collector.clone();
    let a = attrs.clone();
    let memory_rss = meter
        .u64_observable_gauge("iii.worker.memory.rss")
        .with_description("Resident set size in bytes")
        .with_unit("By")
        .with_callback(move |observer| {
            let metrics = c.collect_cached();
            observer.observe(metrics.memory_rss, &a);
        })
        .build();

    let c = collector.clone();
    let a = attrs.clone();
    let memory_virtual = meter
        .u64_observable_gauge("iii.worker.memory.virtual")
        .with_description("Virtual memory in bytes")
        .with_unit("By")
        .with_callback(move |observer| {
            let metrics = c.collect_cached();
            observer.observe(metrics.memory_virtual, &a);
        })
        .build();

    let c = collector.clone();
    let a = attrs.clone();
    let cpu_percent = meter
        .f64_observable_gauge("iii.worker.cpu.percent")
        .with_description("CPU usage percentage")
        .with_unit("%")
        .with_callback(move |observer| {
            let metrics = c.collect_cached();
            observer.observe(metrics.cpu_percent as f64, &a);
        })
        .build();

    let c = collector.clone();
    let a = attrs.clone();
    let uptime = meter
        .f64_observable_gauge("iii.worker.uptime_seconds")
        .with_description("Worker uptime in seconds")
        .with_unit("s")
        .with_callback(move |observer| {
            let metrics = c.collect_cached();
            observer.observe(metrics.uptime_seconds, &a);
        })
        .build();

    WorkerGaugesHandle {
        _memory_rss: memory_rss,
        _memory_virtual: memory_virtual,
        _cpu_percent: cpu_percent,
        _uptime: uptime,
    }
}

/// Handle that keeps the OTEL gauges alive. Drop to stop reporting.
pub struct WorkerGaugesHandle {
    _memory_rss: opentelemetry::metrics::ObservableGauge<u64>,
    _memory_virtual: opentelemetry::metrics::ObservableGauge<u64>,
    _cpu_percent: opentelemetry::metrics::ObservableGauge<f64>,
    _uptime: opentelemetry::metrics::ObservableGauge<f64>,
}
